src/primitive/polyhedron.rs

```diff
@@ -0,0 +1,713 @@
+            bound: vertices
+                .iter()
+                .fold(Aabb3::zero(), |bound, p| bound.grow(p.position)),
+            vertices,
+            edges,
+            faces,
+        }
+    }
+
+    /// Create a new convex polyhedron from the given vertices and faces. Will remove any duplicate
+    /// vertices.
+    pub fn new_with_faces_dedup(
+        vertices: Vec<Point3<S>>,
+        faces: Vec<(usize, usize, usize)>,
+    ) -> Self {
+        let (vertices, map) = dedup_vertices(vertices);
+        Self::new_with_faces(vertices, dedup_faces(faces, map))
+    }
+
+    #[inline]
+    fn brute_force_support_point(&self, direction: Vector3<S>) -> Point3<S> {
+        let (p, _) = self.vertices
+            .iter()
+            .map(|v| (v.position, v.position.dot(direction)))
+            .fold(
+                (Point3::from_value(S::zero()), S::neg_infinity()),
+                |(max_p, max_dot), (v, dot)| if dot > max_dot {
+                    (v.clone(), dot)
+                } else {
+                    (max_p, max_dot)
+                },
+            );
+        p
+    }
+
+    #[inline]
+    fn hill_climb_support_point(&self, direction: Vector3<S>) -> Point3<S> {
+        let mut best_index = 0;
+        let mut best_dot = self.vertices[best_index].position.dot(direction);
+
+        loop {
+            let previous_best_dot = best_dot;
+            let mut edge_index = self.vertices[best_index].edge;
+            let start_edge_index = edge_index;
+
+            loop {
+                let vertex_index = self.edges[edge_index].target_vertex;
+
+                let dot = self.vertices[vertex_index].position.dot(direction);
+                if dot > best_dot {
+                    best_index = vertex_index;
+                    best_dot = dot;
+                }
+
+                edge_index = self.edges[self.edges[edge_index].twin_edge].next_edge;
+                if start_edge_index == edge_index {
+                    break;
+                }
+            }
+
+            if best_dot == previous_best_dot {
+                break;
+            }
+        }
+
+        self.vertices[best_index].position.clone()
+    }
+}
+
+fn dedup_vertices<S>(vertices: Vec<Point3<S>>) -> (Vec<Point3<S>>, HashMap<usize, usize>)
+where
+    S: BaseFloat,
+{
+    let mut vs = Vec::with_capacity(vertices.len() / 2);
+    let mut dup = HashMap::default();
+    for i in 0..vertices.len() {
+        let mut found = false;
+        for j in 0..vs.len() {
+            if vs[j] == vertices[i] {
+                dup.insert(i, j);
+                found = true;
+            }
+        }
+        if !found {
+            vs.push(vertices[i]);
+        }
+    }
+    (vs, dup)
+}
+
+fn dedup_faces(
+    faces: Vec<(usize, usize, usize)>,
+    duplicates: HashMap<usize, usize>,
+) -> Vec<(usize, usize, usize)> {
+    faces
+        .iter()
+        .map(|&(a, b, c)| {
+            (
+                match duplicates.get(&a) {
```

Comment:
could be `*duplicates.get(&a).unwrap_or(&a)`

Reply:
oh, that's clever.